sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala

```diff
@@ -17,285 +17,159 @@
-  def this() = this(null, null) // Required for serialization.
-
-  val currentMin: MutableLiteral = MutableLiteral(null, expr.dataType)
-  val cmp = GreaterThan(currentMin, expr)
-
-  override def update(input: Row): Unit = {
-    if (currentMin.value == null) {
-      currentMin.value = expr.eval(input)
-    } else if(cmp.eval(input) == true) {
-      currentMin.value = expr.eval(input)
-    }
-  }
-
-  override def eval(input: Row): Any = currentMin.value
-}
-
-case class Max(child: Expression) extends PartialAggregate with trees.UnaryNode[Expression] {
-
-  override def nullable: Boolean = true
-  override def dataType: DataType = child.dataType
-  override def toString: String = s"MAX($child)"
-
-  override def asPartial: SplitEvaluation = {
-    val partialMax = Alias(Max(child), "PartialMax")()
-    SplitEvaluation(Max(partialMax.toAttribute), partialMax :: Nil)
-  }
-
-  override def newInstance(): MaxFunction = new MaxFunction(child, this)
-}
-
-case class MaxFunction(expr: Expression, base: AggregateExpression) extends AggregateFunction {
-  def this() = this(null, null) // Required for serialization.
-
-  val currentMax: MutableLiteral = MutableLiteral(null, expr.dataType)
-  val cmp = LessThan(currentMax, expr)
-
-  override def update(input: Row): Unit = {
-    if (currentMax.value == null) {
-      currentMax.value = expr.eval(input)
-    } else if(cmp.eval(input) == true) {
-      currentMax.value = expr.eval(input)
-    }
-  }
+  // Semantically we probably don't need this, however, we need it when
+  // integrating with Hive UDAF(GenericUDAF)
+  @deprecated
+  def terminatePartial(buf: MutableRow): Unit = {}
 
-  override def eval(input: Row): Any = currentMax.value
+  // Output the final result by feeding the aggregation buffer
+  def terminate(input: Row): Any
 }
 
-case class Count(child: Expression) extends PartialAggregate with trees.UnaryNode[Expression] {
+trait AggregateExpression extends Expression with AggregateFunction {
+  self: Product =>
+  type EvaluatedType = Any
 
-  override def nullable: Boolean = false
-  override def dataType: LongType.type = LongType
-  override def toString: String = s"COUNT($child)"
+  var mode: Mode = COMPLETE
 
-  override def asPartial: SplitEvaluation = {
-    val partialCount = Alias(Count(child), "PartialCount")()
-    SplitEvaluation(Coalesce(Seq(Sum(partialCount.toAttribute), Literal(0L))), partialCount :: Nil)
+  def initial(m: Mode): Unit = {
+    this.mode = m
   }
 
-  override def newInstance(): CountFunction = new CountFunction(child, this)
-}
-
-case class CountDistinct(expressions: Seq[Expression]) extends PartialAggregate {
-  def this() = this(null)
+  // Aggregation Buffer data types
+  def bufferDataType: Seq[DataType] = Nil
+  // Is it a distinct aggregate expression?
+  def distinct: Boolean
+  // Is it a distinct like aggregate expression (e.g. Min/Max is distinctLike, while avg is not)
+  def distinctLike: Boolean = false
```

Comment:
I think `distinctLike` is removed in this patch.
That is, all the codes for future optimization need to be removed to simplify this patch.